examples/Calculator.re

```diff
@@ -5,245 +5,341 @@ open Revery.UI.Components;
+let showFloat = float => {
   let string = string_of_float(float);
-  if (String.length(string) > 1
-      && string.[String.length(string) - 1] == '.') {
+  if (String.length(string) > 1 && string.[String.length(string) - 1] == '.') {
     String.sub(string, 0, String.length(string) - 1);
   } else {
     string;
   };
 };
+type state = {
+  operator, /* Operator */
+  result: float, /* The actual numerical result */
+  display: string, /* The equation displayed */
+  number: string /* Current number being typed */
+};
 
-module Calculator = (
-  val component((render, ~children, ()) =>
-    render(() => {
-      /* The equation displayed */
-      let (display, setDisplay) = useState("");
-      /* The actual numerical result */
-      let (result, setResult) = useState(0.);
-      /* Operator */
-      let (op, setOp) = useState(`Nop);
-      /* Current number being typed */
-      let (curNum, setCurNum) = useState("");
-
-      let eval(newOp) {
-        /* Figure out what the string for the next operation will be */
-        let newOpString = switch (newOp) {
-        | `Nop => ""
-        | `Add => "+"
-        | `Sub => "-"
-        | `Mul => "×"
-        | `Div => "÷"
-        };
-        /* Split the current display on ! and get everything after (to clear errors) */
-        let partitionedDisplay = String.split_on_char('!', display);
-        let display = List.nth(partitionedDisplay, List.length(partitionedDisplay) - 1);
-        let (newDisplay, newResult) = switch (op) {
-        | #operator when curNum == "" =>
-          ("Error: Can't evaluate binary operator without input!", result)
-        | `Nop => (curNum ++ newOpString, float_of_string(curNum))
-        | `Add => (display ++ curNum ++ newOpString, result +. float_of_string(curNum))
-        | `Sub => (display ++ curNum ++ newOpString, result -. float_of_string(curNum))
-        | `Mul => (display ++ curNum ++ newOpString, result *. float_of_string(curNum))
-        | `Div =>
-          if (float_of_string(curNum) != 0.) {
-            (display ++ curNum ++ newOpString, result /. float_of_string(curNum));
-          } else {
-            ("Error: Divide by zero!", result);
-          }
-        };
-        setResult(newResult);
-        setDisplay(newDisplay);
-        newResult;
-      };
-
-      let perform(newOp) {
-        let _newResult = eval(newOp);
-        setOp(newOp);
-        setCurNum("");
-      };
-
-      let calculate() {
-        let newResult = eval(`Nop);
-        setOp(`Nop);
-        setCurNum(showFloat(newResult));
-      };
+type action =
+  | BackspaceKeyPressed
+  | ClearKeyPressed(bool) /*AC pressed */
+  | DotKeyPressed
+  | NumberKeyPressed(string)
+  | OperationKeyPressed(operator)
+  | PlusMinusKeyPressed
+  | ResultKeyPressed;
 
-      let clear(~clearAcc) {
-        if (clearAcc) {
-          setOp(`Nop);
-          setResult(0.);
-          setDisplay("");
-        };
-        setCurNum("");
-      };
+let eval = (state, newOp) => {
```

Comment:
Now that we have the reducer, maybe we can refactor `eval` into an action? Thoughts on this?

Reply:
I think `eval` wouldn't really fit an action-like role. My mental model for actions is "actions are like newspapers", in the sense that actions need to tell you something that happened in your app (generally user originated, or network originated). This talk has a really great overview on what are good candidates for actions: https://youtu.be/Dm9NgjR5Jn4?t=485

`eval` would fit more a "command" role. Because multiple actions (in this case 2) need to perform the evaluation, I think it makes sense to keep `eval` as just a helper function that encapsulates the common logic in the two actions reducers. There are plenty of ways to solve this though 😄 Do you think we should rethink the way `eval` and the current actions fit together?